ibft/storage/store.go

```diff
@@ -38,46 +41,137 @@ func init() {
 	}
 }
 
-// ibftStorage struct
+// participantStorage struct
 // instanceType is what separates different iBFT eth2 duty types (attestation, proposal and aggregation)
-type ibftStorage struct {
+type participantStorage struct {
 	prefix         []byte
+	oldPrefix      string // kept back for cleanup
 	db             basedb.Database
 	participantsMu sync.Mutex
 }
 
-// New create new ibft storage
-func New(db basedb.Database, prefix string) qbftstorage.QBFTStore {
-	return &ibftStorage{
-		prefix: []byte(prefix),
-		db:     db,
+// New create new participant store
+func New(db basedb.Database, prefix spectypes.BeaconRole) qbftstorage.ParticipantStore {
+	role := byte(prefix & 0xff)
+	return &participantStorage{
+		prefix:    []byte{role},
+		oldPrefix: prefix.String(),
+		db:        db,
 	}
 }
 
-// CleanAllInstances removes all StoredInstance's & highest StoredInstance's for msgID.
-func (i *ibftStorage) CleanAllInstances(msgID []byte) error {
-	prefix := i.prefix
-	prefix = append(prefix, msgID[:]...)
-	prefix = append(prefix, []byte(instanceKey)...)
-	err := i.db.DropPrefix(prefix)
+func (i *participantStorage) StartCleanupJob(ctx context.Context, logger *zap.Logger, slotTickerProvider slotticker.Provider, retain int) {
+	logger.Debug("start stale slot cleanup job", zap.Int("retain", retain))
+	ticker := slotTickerProvider()
+	<-ticker.Next()
+	threashold := ticker.Slot() - phase0.Slot(retain) // #nosec G115
+
+	// on start we remove ALL slots below the threashold
+	start := time.Now()
+	count := i.removeSlotsOlderThan(logger, threashold)
+
+	logger.Info("removed stale slot entries", zap.String("store", i.ID()), zap.Int("count", count), zap.Duration("took", time.Since(start)))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.Next():
+			threashold := ticker.Slot() - phase0.Slot(retain) - 1 // #nosec G115
+			if err := i.removeSlotAt(threashold); err != nil {
+				logger.Error("remove slot at", fields.Slot(threashold))
+			}
+		}
+	}
+}
+
+// removes ALL entries that have given slot in their prefix
+func (i *participantStorage) removeSlotAt(slot phase0.Slot) error {
+	var keySet [][]byte
+
+	prefix := i.makePrefix(slotToByteSlice(slot))
+
+	tx := i.db.Begin()
+	defer tx.Discard()
+
+	// filter and collect keys
+	err := i.db.UsingReader(tx).GetAll(prefix, func(i int, o basedb.Obj) error {
+		keySet = append(keySet, o.Key)
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("collect keys of stale slots: %w", err)
+	}
+
+	for _, id := range keySet {
+		if err := i.db.Using(tx).Delete(append(prefix, id...), nil); err != nil {
+			return fmt.Errorf("remove slot: %w", err)
```

Comment:
couldn't use `DropPrefix` here because it doesn't support transactions; and 4 goroutines dropping prefix concomitantly caused database errors.